backend/infrahub/services/__init__.py

```diff
@@ -28,23 +33,77 @@ def __init__(
         client: Optional[InfrahubClient] = None,
         database: Optional[InfrahubDatabase] = None,
         message_bus: Optional[InfrahubMessageBus] = None,
-        http: InfrahubHTTP | None = None,
         workflow: Optional[InfrahubWorkflow] = None,
         event: InfrahubEventService | None = None,
         log: Optional[InfrahubLogger] = None,
         component_type: Optional[ComponentType] = None,
     ) -> None:
-        self.cache = cache or InfrahubCache()
+        self._cache = cache
         self._client = client
         self._database = database
-        self.message_bus = message_bus or InfrahubMessageBus()
-        self.workflow = workflow or InfrahubWorkflow()
+        self._message_bus = message_bus
         self.event = event or InfrahubEventService()
         self.log = log or get_logger()
         self.component_type = component_type or ComponentType.NONE
-        self.http = http or HttpxAdapter()
+        self.http = HttpxAdapter()
         self.scheduler = InfrahubScheduler()
         self.component = InfrahubComponent()
+        self._workflow = workflow
+
+        # Hack for testing purposes. WorkflowLocalExecution needs a reference to services within execute_workflow
+        # and we don't want to call workflow.initialize within each of our tests.
+        if isinstance(self._workflow, WorkflowLocalExecution):
+            self._workflow.service = self
+
+    @classmethod
+    async def init_and_initialize(
```

Comment:
Do we need both `initialize` and `init_and_initialize`.? could we merge both into `initialize` or `init` ?

Reply:
Some tests currently instantiate an `InfrahubServices` object without calling `initialize` and break if we initialize them, typically because `InfrahubComponent` is always instantiated and requires an `InfrahubCache` that might not be there. I am realizing that solution here would be to have `InfrahubComponent` explicitly relying on `InfrahubCache`, and so we instantiate it only if a cache is present (which brings us back to [this discussion](https://github.com/opsmill/infrahub/pull/5421#discussion_r1909549939)). I will have a look at that.